### PR TITLE
[REFACTOR][NODE][RUNTIME] Move Node to the new Object protocol.

### PR DESCRIPTION
**Pull request by tqchen**
This PR removes the original node system, and make node as a subclass of Object.
  This is a major refactor towards a better unified runtime object system, see RFC #4116 

  List of changes in the refactor:
    
  - We now hide data_ field, use Downcast explicitly to get a sub-class object.
  - Removed the node system FFI in python.
  - Removed the node C API, instead use PackedFunc for list and get attrs.
  - Change relay::Op::set_attr_type_key(attr_key_name) to relay::Op::set_attr_type<AttrType>().
      - This change was necessary because of the new Object registration mechanism.
      - Subsequent changes to the op registrations
      - The change revealed a few previous problems that is now fixed.
  - Patched up a few missing node type registration.
      - Now we will raise an error if we register object that is not registered.
  - The original node.h and container.h are kept in the same location.
  - Calling convention: kObjectHandle now equals the old kNodeHandle, kNodeHandle is removed.
  - IRFunctor now dispatches on ObjectRef.
  - Update to the new type checking API: is_type, derived_from are replaced by IsInstance.
  - Removed .hash member function, instead use C++ convention hasher functors.

### CODE REVIEW

**Comment by tqchen** [comment]
@icemelon9 @jroesch please help to take a look

File: include/tvm/base.h [hunk 1]
```diff
@@ -47,9 +47,10 @@ using ::tvm::AttrVisitor;
  */
 #define TVM_DEFINE_NODE_REF_METHODS(TypeName, BaseTypeName, NodeName)   \
   TypeName() {}                                                         \
-  explicit TypeName(::tvm::NodePtr<::tvm::Node> n) : BaseTypeName(n) {} \
+  explicit TypeName(::tvm::ObjectPtr<::tvm::Object> n)                  \
+      : BaseTypeName(n) {}                                              \
   const NodeName* operator->() const {                                  \
-    return static_cast<const NodeName*>(node_.get());                   \
+    return static_cast<const NodeName*>(data_.get());                   \
```
Comment (zhiics) [inline include/tvm/base.h]: get()

File: include/tvm/base.h [hunk 2]
```diff
@@ -75,12 +76,12 @@ using ::tvm::AttrVisitor;
...
-      CHECK(node_ != nullptr);                                          \
-      if (!node_.unique())  {                                           \
+      CHECK(data_ != nullptr);                                          \
+      if (!data_.unique())  {                                           \
         NodePtr<NodeName> n = make_node<NodeName>(*(operator->()));     \
-        NodePtr<Node>(std::move(n)).swap(node_);                        \
+        ObjectPtr<Object>(std::move(n)).swap(data_);                    \
       }                                                                 \
-      return static_cast<NodeName*>(node_.get());                       \
+      return static_cast<NodeName*>(data_.get());                       \
```
Comment (zhiics) [inline include/tvm/base.h]: get_mutable()

File: include/tvm/expr.h [hunk 3]
```diff
@@ -197,21 +197,21 @@ class Integer : public Expr {
...
+    data_ = other.data_;
     return *this;
   }
   /*!
    * \brief Get pointer to the internal value.
    * \return the content of the integer.
    */
   const IntImm* operator->() const {
-    return static_cast<const IntImm*>(node_.get());
+    return static_cast<const IntImm*>(data_.get());
```
Comment (zhiics) [inline include/tvm/expr.h]: get()

File: include/tvm/expr.h [hunk 4]
```diff
@@ -423,7 +423,7 @@ class IterVarNode : public Node {
 
 // inline implementations
 inline const IterVarNode* IterVar::operator->() const {
-  return static_cast<const IterVarNode*>(node_.get());
+  return static_cast<const IterVarNode*>(data_.get());
```
Comment (zhiics) [inline include/tvm/expr.h]: get()

**Comment by tqchen** [comment]
Thanks @zhiics i have addressed your comments

**Comment by tqchen** [comment]
Thanks @zhiics

**Pull request merged**